{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<div>\n",
    "    <table style=\"border:2px solid white;\" cellspacing=\"0\" cellpadding=\"0\" border-collapse: collapse; border-spacing: 0;>\n",
    "      <tr> \n",
    "        <th style=\"background-color:white\"> <img src=\"../media/ccal-logo-D3.png\" width=225 height=225></th>\n",
    "        <th style=\"background-color:white\"> <img src=\"../media/logoMoores.jpg\" width=175 height=175></th>\n",
    "        <th style=\"background-color:white\"> <img src=\"../media/GP.png\" width=200 height=200></th>\n",
    "        <th style=\"background-color:white\"> <img src=\"../media/UCSD_School_of_Medicine_logo.png\" width=175 height=175></th> \n",
    "        <th style=\"background-color:white\"> <img src=\"../media/Broad.png\" width=130 height=130></th> \n",
    "      </tr>\n",
    "    </table>\n",
    "</div>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<hr style=\"border: none; border-bottom: 3px solid #88BBEE;\">\n",
    "# **Onco-*GPS* Methodology**\n",
    "## **Chapter 4. Annotating the Transcriptional Components**"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<div>\n",
    "    <img style=\"float: left\" src=\"../media/authors.png\" width=800 height=40>\n",
    "</div>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Date:** April 17, 2017\n",
    "\n",
    "\n",
    "**Article:** [*Kim et al.* Decomposing Oncogenic Transcriptional Signatures to Generate Maps of Divergent Cellular States](https://drive.google.com/file/d/0B0MQqMWLrsA4b2RUTTAzNjFmVkk/view?usp=sharing)\n",
    "\n",
    "**Analysis overview:** In this chapter we perform a detailed analysis of the KRAS transcriptional components produced by the NMF decomposition in chapter 3 in order to assign a biological interpretation to each component. \n",
    "\n",
    "<div>\n",
    "    <img src=\"../media/method_chap3.png\" width=2144 height=1041>\n",
    "</div>\n",
    "\n",
    "The analysis consists of the following steps:\n",
    "* Define a target profile for each component in the CCLE Reference Dataset using the amplitudes of the $H$ matrix. This matrix represents the magnitude of each NMF component per sample. \n",
    "* Using the Information Coefficient (IC) ([*Kim, J.W., Botvinnik 2016*](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4868596/)) to estimate the degree of association of each component target profile and several types of genomic features.\n",
    "\n",
    "The genomic features include the following:\n",
    "\n",
    "1. **Mutations and Copy Number Alterations (CNA).** CCLE mutation and copy number datasets (www.broadinstitute.org/ccle, [*Barretina et al. 2012*](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3320027/)).\n",
    "2.\t **Gene expression.** CCLE RNA Seq dataset (http://www.broadinstitute.org/ccle, [*Barretina et al. 2012*](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3320027/)).\n",
    "3.\t  **Pathway expression** (single sample GSEA of MSigDB gene sets) MSigDB v5.1 sub-collections c2, c5, c6 and h www.msigdb.org, (Liberzon et al. 2011; [*Liberzon et al. 2016. Cell Systems, 1(6), pp.417–425.*](https://www.ncbi.nlm.nih.gov/pubmed/26771021). and a few additional gene sets (see supplementary information in the article).\n",
    "4.\t**Transcription factors and master regulators expression** (single sample GSEA of gene sets) MSigDB v5.1, ([*Liberzon et al. 2011*](https://www.ncbi.nlm.nih.gov/pubmed/21546393)) http://www.msigdb.org, sub-collection c3 and 1,598 IPA gene sets, http://www.ingenuity.com.\n",
    "5.\t **Protein expression.** CCLE Reverse Phased Protein Array (RPPA) dataset (http://www.broadinstitute.org/ccle, [*Barretina et al. 2012*](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3320027/)).\n",
    "6.\t **Drug sensitivity.** CCLE dataset (http://www.broadinstitute.org/ccle, [*Barretina et al. 2012*](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3320027/))\n",
    "7.\t**Gene dependency.** RNAi Achilles dataset, http://www.broadinstitute.org/achilles, ([*Cowley et al. 2014*](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4432652/)).\n",
    "\n",
    "Go to the [next chapter (5)](5 Defining Cellular States and Generating Onco-GPS Map.ipynb).\n",
    "Back to the [introduction chapter (0)](0 Introduction and Overview.ipynb)."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<hr style=\"border: none; border-bottom: 3px solid #88BBEE;\">\n",
    "### 1. Set up notebook and import Computational Cancer Analysis Library ([CCAL](https://github.com/KwatME/ccal))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2017-07-09T19:55:27.334846Z",
     "start_time": "2017-07-09T19:55:25.028168"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The autoreload extension is already loaded. To reload it, use:\n",
      "  %reload_ext autoreload\n"
     ]
    }
   ],
   "source": [
    "from environment import *\n",
    "\n",
    "%matplotlib inline\n",
    "%load_ext autoreload\n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 2. Read the annotation datasets table and $H$  matrix"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The table describing the datasets that will be used in the annotation analysis (annotation.data_table.txt) is included in the directory \"../data.\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Data Name</th>\n",
       "      <th>Data Type</th>\n",
       "      <th>Emphasis</th>\n",
       "      <th>Filepath</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>drug_sensitivity</td>\n",
       "      <td>continuous</td>\n",
       "      <td>low</td>\n",
       "      <td>../data/ccle_drug_sensitivity.gct</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>gene_expression</td>\n",
       "      <td>continuous</td>\n",
       "      <td>high</td>\n",
       "      <td>../data/ccle_gene_expression.gct</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>gene_dependency</td>\n",
       "      <td>continuous</td>\n",
       "      <td>low</td>\n",
       "      <td>../data/ExpandedGeneZSolsCleaned.gct</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>mutation</td>\n",
       "      <td>binary</td>\n",
       "      <td>high</td>\n",
       "      <td>../data/ccle_mut_CNA.gct</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>pathway_expression</td>\n",
       "      <td>continuous</td>\n",
       "      <td>high</td>\n",
       "      <td>../data/ccle_pathway_expression_all.gct</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>protein_expression</td>\n",
       "      <td>continuous</td>\n",
       "      <td>high</td>\n",
       "      <td>../data/ccle_protein_expression.gct</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>regulator</td>\n",
       "      <td>continuous</td>\n",
       "      <td>high</td>\n",
       "      <td>../data/ccle_regulator.gct</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>tissue</td>\n",
       "      <td>binary</td>\n",
       "      <td>high</td>\n",
       "      <td>../data/ccle_tissue.gct</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            Data Name   Data Type Emphasis  \\\n",
       "0    drug_sensitivity  continuous      low   \n",
       "1     gene_expression  continuous     high   \n",
       "2     gene_dependency  continuous      low   \n",
       "3            mutation      binary     high   \n",
       "4  pathway_expression  continuous     high   \n",
       "5  protein_expression  continuous     high   \n",
       "6           regulator  continuous     high   \n",
       "7              tissue      binary     high   \n",
       "\n",
       "                                  Filepath  \n",
       "0        ../data/ccle_drug_sensitivity.gct  \n",
       "1         ../data/ccle_gene_expression.gct  \n",
       "2     ../data/ExpandedGeneZSolsCleaned.gct  \n",
       "3                 ../data/ccle_mut_CNA.gct  \n",
       "4  ../data/ccle_pathway_expression_all.gct  \n",
       "5      ../data/ccle_protein_expression.gct  \n",
       "6               ../data/ccle_regulator.gct  \n",
       "7                  ../data/ccle_tissue.gct  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.read_csv('../data/annotation.data_table.txt', sep='\\t')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This function below will read that table and the datasets defined in the \"Filepath\" column"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Making data bundle for drug_sensitivity ...\n",
      "\tLoaded ../data/ccle_drug_sensitivity.gct.\n",
      "Making data bundle for gene_expression ...\n",
      "\tLoaded ../data/ccle_gene_expression.gct.\n",
      "Making data bundle for gene_dependency ...\n",
      "\tLoaded ../data/ExpandedGeneZSolsCleaned.gct.\n",
      "Making data bundle for mutation ...\n",
      "\tLoaded ../data/ccle_mut_CNA.gct.\n",
      "Making data bundle for pathway_expression ...\n",
      "\tLoaded ../data/ccle_pathway_expression_all.gct.\n",
      "Making data bundle for protein_expression ...\n",
      "\tLoaded ../data/ccle_protein_expression.gct.\n",
      "Making data bundle for regulator ...\n",
      "\tLoaded ../data/ccle_regulator.gct.\n",
      "Making data bundle for tissue ...\n",
      "\tLoaded ../data/ccle_tissue.gct.\n"
     ]
    }
   ],
   "source": [
    "data_table = ccal.load_data_table('../data/annotation.data_table.txt')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    " Read the $H$ matrix produced in notebook 3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "h_matrix = ccal.read_gct('../results/nmf_cc/nmf/nmf_k9_h.gct')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>A101D_SKIN</th>\n",
       "      <th>A172_CENTRAL_NERVOUS_SYSTEM</th>\n",
       "      <th>A204_SOFT_TISSUE</th>\n",
       "      <th>A2058_SKIN</th>\n",
       "      <th>A2780_OVARY</th>\n",
       "      <th>A375_SKIN</th>\n",
       "      <th>A498_KIDNEY</th>\n",
       "      <th>A549_LUNG</th>\n",
       "      <th>A673_BONE</th>\n",
       "      <th>A704_KIDNEY</th>\n",
       "      <th>...</th>\n",
       "      <th>WM88_SKIN</th>\n",
       "      <th>WM983B_SKIN</th>\n",
       "      <th>YAPC_PANCREAS</th>\n",
       "      <th>YD10B_UPPER_AERODIGESTIVE_TRACT</th>\n",
       "      <th>YD38_UPPER_AERODIGESTIVE_TRACT</th>\n",
       "      <th>YD8_UPPER_AERODIGESTIVE_TRACT</th>\n",
       "      <th>YH13_CENTRAL_NERVOUS_SYSTEM</th>\n",
       "      <th>YKG1_CENTRAL_NERVOUS_SYSTEM</th>\n",
       "      <th>ZR751_BREAST</th>\n",
       "      <th>ZR7530_BREAST</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>C1</th>\n",
       "      <td>87.390561</td>\n",
       "      <td>93.350830</td>\n",
       "      <td>393.899758</td>\n",
       "      <td>255.625865</td>\n",
       "      <td>12.230046</td>\n",
       "      <td>125.904286</td>\n",
       "      <td>15.489376</td>\n",
       "      <td>5.201425e-08</td>\n",
       "      <td>0.163371</td>\n",
       "      <td>0.001205</td>\n",
       "      <td>...</td>\n",
       "      <td>415.948914</td>\n",
       "      <td>305.351132</td>\n",
       "      <td>2248.563800</td>\n",
       "      <td>685.682262</td>\n",
       "      <td>1798.525914</td>\n",
       "      <td>104.278153</td>\n",
       "      <td>199.214356</td>\n",
       "      <td>3.819411e-01</td>\n",
       "      <td>4721.947367</td>\n",
       "      <td>4957.456757</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C3</th>\n",
       "      <td>518.647456</td>\n",
       "      <td>7.308027</td>\n",
       "      <td>199.238501</td>\n",
       "      <td>176.804193</td>\n",
       "      <td>545.601644</td>\n",
       "      <td>391.325939</td>\n",
       "      <td>452.722609</td>\n",
       "      <td>8.167350e+02</td>\n",
       "      <td>444.891880</td>\n",
       "      <td>1664.636732</td>\n",
       "      <td>...</td>\n",
       "      <td>295.629218</td>\n",
       "      <td>559.708338</td>\n",
       "      <td>2155.739326</td>\n",
       "      <td>355.439796</td>\n",
       "      <td>302.706434</td>\n",
       "      <td>0.000576</td>\n",
       "      <td>328.560693</td>\n",
       "      <td>1.112408e+02</td>\n",
       "      <td>357.322949</td>\n",
       "      <td>653.472173</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C9</th>\n",
       "      <td>11.208338</td>\n",
       "      <td>622.861218</td>\n",
       "      <td>717.757144</td>\n",
       "      <td>3.069694</td>\n",
       "      <td>446.187704</td>\n",
       "      <td>0.000439</td>\n",
       "      <td>740.861040</td>\n",
       "      <td>6.040093e+02</td>\n",
       "      <td>975.677929</td>\n",
       "      <td>1.341833</td>\n",
       "      <td>...</td>\n",
       "      <td>463.865834</td>\n",
       "      <td>118.997647</td>\n",
       "      <td>1890.725829</td>\n",
       "      <td>5669.601718</td>\n",
       "      <td>3901.938851</td>\n",
       "      <td>3309.332406</td>\n",
       "      <td>600.933737</td>\n",
       "      <td>1.969634e-10</td>\n",
       "      <td>457.725732</td>\n",
       "      <td>272.206384</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C8</th>\n",
       "      <td>239.003422</td>\n",
       "      <td>479.108337</td>\n",
       "      <td>693.482964</td>\n",
       "      <td>478.286114</td>\n",
       "      <td>1696.595193</td>\n",
       "      <td>503.049222</td>\n",
       "      <td>269.120017</td>\n",
       "      <td>6.510322e+02</td>\n",
       "      <td>2021.718516</td>\n",
       "      <td>372.320844</td>\n",
       "      <td>...</td>\n",
       "      <td>363.270689</td>\n",
       "      <td>482.473882</td>\n",
       "      <td>14.761625</td>\n",
       "      <td>512.785549</td>\n",
       "      <td>401.354229</td>\n",
       "      <td>23.818909</td>\n",
       "      <td>445.879251</td>\n",
       "      <td>9.487706e+02</td>\n",
       "      <td>942.540098</td>\n",
       "      <td>814.310434</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C6</th>\n",
       "      <td>7166.174137</td>\n",
       "      <td>1208.779409</td>\n",
       "      <td>1774.414891</td>\n",
       "      <td>6276.135642</td>\n",
       "      <td>1362.720325</td>\n",
       "      <td>4779.659070</td>\n",
       "      <td>838.082289</td>\n",
       "      <td>8.212941e-03</td>\n",
       "      <td>1611.742127</td>\n",
       "      <td>1417.850903</td>\n",
       "      <td>...</td>\n",
       "      <td>7112.681988</td>\n",
       "      <td>7116.940554</td>\n",
       "      <td>543.418521</td>\n",
       "      <td>209.558478</td>\n",
       "      <td>493.361221</td>\n",
       "      <td>24.717198</td>\n",
       "      <td>1032.356299</td>\n",
       "      <td>2.382286e+03</td>\n",
       "      <td>285.983913</td>\n",
       "      <td>146.797171</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C7</th>\n",
       "      <td>1016.009451</td>\n",
       "      <td>1672.966975</td>\n",
       "      <td>433.423134</td>\n",
       "      <td>955.066087</td>\n",
       "      <td>0.003192</td>\n",
       "      <td>1688.565623</td>\n",
       "      <td>961.965537</td>\n",
       "      <td>1.294276e+03</td>\n",
       "      <td>45.169949</td>\n",
       "      <td>28.547109</td>\n",
       "      <td>...</td>\n",
       "      <td>156.910209</td>\n",
       "      <td>617.164602</td>\n",
       "      <td>2113.148083</td>\n",
       "      <td>1606.060845</td>\n",
       "      <td>2266.279767</td>\n",
       "      <td>1820.623602</td>\n",
       "      <td>1502.274660</td>\n",
       "      <td>1.324849e+03</td>\n",
       "      <td>0.016081</td>\n",
       "      <td>0.001204</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C5</th>\n",
       "      <td>764.479691</td>\n",
       "      <td>725.646766</td>\n",
       "      <td>517.555359</td>\n",
       "      <td>459.776995</td>\n",
       "      <td>0.000010</td>\n",
       "      <td>0.112397</td>\n",
       "      <td>4271.867752</td>\n",
       "      <td>3.473658e+03</td>\n",
       "      <td>23.952947</td>\n",
       "      <td>5576.769071</td>\n",
       "      <td>...</td>\n",
       "      <td>952.997119</td>\n",
       "      <td>568.321975</td>\n",
       "      <td>719.250074</td>\n",
       "      <td>2.775659</td>\n",
       "      <td>541.816602</td>\n",
       "      <td>1600.013238</td>\n",
       "      <td>509.370896</td>\n",
       "      <td>2.454880e+02</td>\n",
       "      <td>505.321577</td>\n",
       "      <td>310.379079</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C2</th>\n",
       "      <td>64.829650</td>\n",
       "      <td>1569.836315</td>\n",
       "      <td>3713.903057</td>\n",
       "      <td>1189.301135</td>\n",
       "      <td>4612.165638</td>\n",
       "      <td>1692.563251</td>\n",
       "      <td>1041.812928</td>\n",
       "      <td>2.293922e+03</td>\n",
       "      <td>2825.029588</td>\n",
       "      <td>777.419387</td>\n",
       "      <td>...</td>\n",
       "      <td>737.636509</td>\n",
       "      <td>493.994389</td>\n",
       "      <td>220.968519</td>\n",
       "      <td>436.847693</td>\n",
       "      <td>112.120030</td>\n",
       "      <td>474.166240</td>\n",
       "      <td>277.577697</td>\n",
       "      <td>1.885415e+03</td>\n",
       "      <td>1270.077326</td>\n",
       "      <td>1061.667687</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C4</th>\n",
       "      <td>1048.429865</td>\n",
       "      <td>3514.982837</td>\n",
       "      <td>1599.533720</td>\n",
       "      <td>949.324498</td>\n",
       "      <td>1395.774975</td>\n",
       "      <td>1317.373754</td>\n",
       "      <td>1505.001742</td>\n",
       "      <td>8.602564e+02</td>\n",
       "      <td>2186.326459</td>\n",
       "      <td>488.753903</td>\n",
       "      <td>...</td>\n",
       "      <td>427.337725</td>\n",
       "      <td>656.160837</td>\n",
       "      <td>0.500336</td>\n",
       "      <td>580.722452</td>\n",
       "      <td>129.765634</td>\n",
       "      <td>2539.073702</td>\n",
       "      <td>4918.553670</td>\n",
       "      <td>3.186260e+03</td>\n",
       "      <td>940.169658</td>\n",
       "      <td>1203.297652</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>9 rows × 750 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     A101D_SKIN  A172_CENTRAL_NERVOUS_SYSTEM  A204_SOFT_TISSUE   A2058_SKIN  \\\n",
       "C1    87.390561                    93.350830        393.899758   255.625865   \n",
       "C3   518.647456                     7.308027        199.238501   176.804193   \n",
       "C9    11.208338                   622.861218        717.757144     3.069694   \n",
       "C8   239.003422                   479.108337        693.482964   478.286114   \n",
       "C6  7166.174137                  1208.779409       1774.414891  6276.135642   \n",
       "C7  1016.009451                  1672.966975        433.423134   955.066087   \n",
       "C5   764.479691                   725.646766        517.555359   459.776995   \n",
       "C2    64.829650                  1569.836315       3713.903057  1189.301135   \n",
       "C4  1048.429865                  3514.982837       1599.533720   949.324498   \n",
       "\n",
       "    A2780_OVARY    A375_SKIN  A498_KIDNEY     A549_LUNG    A673_BONE  \\\n",
       "C1    12.230046   125.904286    15.489376  5.201425e-08     0.163371   \n",
       "C3   545.601644   391.325939   452.722609  8.167350e+02   444.891880   \n",
       "C9   446.187704     0.000439   740.861040  6.040093e+02   975.677929   \n",
       "C8  1696.595193   503.049222   269.120017  6.510322e+02  2021.718516   \n",
       "C6  1362.720325  4779.659070   838.082289  8.212941e-03  1611.742127   \n",
       "C7     0.003192  1688.565623   961.965537  1.294276e+03    45.169949   \n",
       "C5     0.000010     0.112397  4271.867752  3.473658e+03    23.952947   \n",
       "C2  4612.165638  1692.563251  1041.812928  2.293922e+03  2825.029588   \n",
       "C4  1395.774975  1317.373754  1505.001742  8.602564e+02  2186.326459   \n",
       "\n",
       "    A704_KIDNEY      ...          WM88_SKIN  WM983B_SKIN  YAPC_PANCREAS  \\\n",
       "C1     0.001205      ...         415.948914   305.351132    2248.563800   \n",
       "C3  1664.636732      ...         295.629218   559.708338    2155.739326   \n",
       "C9     1.341833      ...         463.865834   118.997647    1890.725829   \n",
       "C8   372.320844      ...         363.270689   482.473882      14.761625   \n",
       "C6  1417.850903      ...        7112.681988  7116.940554     543.418521   \n",
       "C7    28.547109      ...         156.910209   617.164602    2113.148083   \n",
       "C5  5576.769071      ...         952.997119   568.321975     719.250074   \n",
       "C2   777.419387      ...         737.636509   493.994389     220.968519   \n",
       "C4   488.753903      ...         427.337725   656.160837       0.500336   \n",
       "\n",
       "    YD10B_UPPER_AERODIGESTIVE_TRACT  YD38_UPPER_AERODIGESTIVE_TRACT  \\\n",
       "C1                       685.682262                     1798.525914   \n",
       "C3                       355.439796                      302.706434   \n",
       "C9                      5669.601718                     3901.938851   \n",
       "C8                       512.785549                      401.354229   \n",
       "C6                       209.558478                      493.361221   \n",
       "C7                      1606.060845                     2266.279767   \n",
       "C5                         2.775659                      541.816602   \n",
       "C2                       436.847693                      112.120030   \n",
       "C4                       580.722452                      129.765634   \n",
       "\n",
       "    YD8_UPPER_AERODIGESTIVE_TRACT  YH13_CENTRAL_NERVOUS_SYSTEM  \\\n",
       "C1                     104.278153                   199.214356   \n",
       "C3                       0.000576                   328.560693   \n",
       "C9                    3309.332406                   600.933737   \n",
       "C8                      23.818909                   445.879251   \n",
       "C6                      24.717198                  1032.356299   \n",
       "C7                    1820.623602                  1502.274660   \n",
       "C5                    1600.013238                   509.370896   \n",
       "C2                     474.166240                   277.577697   \n",
       "C4                    2539.073702                  4918.553670   \n",
       "\n",
       "    YKG1_CENTRAL_NERVOUS_SYSTEM  ZR751_BREAST  ZR7530_BREAST  \n",
       "C1                 3.819411e-01   4721.947367    4957.456757  \n",
       "C3                 1.112408e+02    357.322949     653.472173  \n",
       "C9                 1.969634e-10    457.725732     272.206384  \n",
       "C8                 9.487706e+02    942.540098     814.310434  \n",
       "C6                 2.382286e+03    285.983913     146.797171  \n",
       "C7                 1.324849e+03      0.016081       0.001204  \n",
       "C5                 2.454880e+02    505.321577     310.379079  \n",
       "C2                 1.885415e+03   1270.077326    1061.667687  \n",
       "C4                 3.186260e+03    940.169658    1203.297652  \n",
       "\n",
       "[9 rows x 750 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "h_matrix"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 3. Find the top genomic features that match each component profile\n",
    "The annotation consists of running the association analysis for each component against all the genomic datasets. Because this a double iteration over componets and feature datasets it will take hours to complete. As the program runs it will display the specific target vs. features comparison being made."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "ccal.association.make_association_panels(\n",
    "    target=h_matrix,\n",
    "    data_bundle=data_table,\n",
    "    dropna='all',\n",
    "    target_ascending=False,\n",
    "    target_prefix='',\n",
    "    data_prefix='',\n",
    "    target_type='continuous',\n",
    "    n_jobs=1,\n",
    "    n_features=20,\n",
    "    n_samplings=30,\n",
    "    n_permutations=50,\n",
    "    random_seed=12345,\n",
    "    directory_path='../results/component_annotation')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 4. Make summary panels for selected component associations\n",
    "\n",
    "Here we will select specific sets of top scoring features of particular biological interest for many of the components\n",
    "\n",
    "#### 4.1  Selected associations for component C3 \n",
    "\n",
    "The results of the association analysis shows that component C3 is the most associated with KRAS mutation status. As can be seen below in addition to KRAS mutation status, C3 is also associated with a KRAS dependency signature and with a profile of KRAS RNAi dependency from Project Achilles. These findings suggest that the transcriptional activity of KRAS represented by C3 underlies the KRAS dependence phenotype [*Singh et al. 2009*](https://www.ncbi.nlm.nih.gov/pubmed/19690457) and helps to explain the fact that KRAS mutant cancers with low C3 scores, i.e., samples on the right side of the heatmap below do not depend on KRAS for their survival."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Selected KRAS-Related Features for KRAS Component C3 (Figure 4A)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Select component\n",
    "component = 'C3'\n",
    "target = h_matrix.ix[component, :]\n",
    "target.name = 'KRAS Component C3'\n",
    "\n",
    "# Select specific annotation features\n",
    "annotation_features = {\n",
    "    'mutation': {\n",
    "        'index': ['KRAS_MUT'],\n",
    "        'alias': ['KRAS mut']\n",
    "    },\n",
    "    'pathway_expression': {\n",
    "        'index': ['SINGH_KRAS_DEPENDENCY_SIGNATURE_'],\n",
    "        'alias': ['KRAS Dependency']\n",
    "    },\n",
    "    'gene_dependency': {\n",
    "        'index': ['KRAS'],\n",
    "        'alias': ['KRAS']\n",
    "    }\n",
    "}\n",
    "\n",
    "# Load annotaion bundle\n",
    "annotation_bundle = ccal.support.file.load_data_table(\n",
    "    '../data/annotation.data_table.txt', annotation_features)\n",
    "\n",
    "# Load annotation files\n",
    "annotation_files = {\n",
    "    'mutation':\n",
    "    '../results/component_annotation/c3_vs_mutation.txt',\n",
    "    'pathway_expression':\n",
    "    '../results/component_annotation/c3_vs_pathway_expression.txt',\n",
    "    'gene_dependency':\n",
    "    '../results/component_annotation/c3_vs_gene_dependency.txt'\n",
    "}\n",
    "\n",
    "# Make summary panels\n",
    "ccal.association.make_association_summary_panel(\n",
    "    target,\n",
    "    annotation_bundle,\n",
    "    annotation_files,\n",
    "    order=['mutation', 'pathway_expression', 'gene_dependency'],\n",
    "    title='Selected KRAS-Related Features for KRAS Component C3',\n",
    "    filepath='../results/C3.WNT.vignette.pdf')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Selected WNT-Related Features for KRAS Component C3 (Figure 4E)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Select component\n",
    "component = 'C3'\n",
    "target = h_matrix.ix[component, :]\n",
    "target.name = 'KRAS Component C3'\n",
    "\n",
    "# Select specific annotation features\n",
    "annotation_features = {\n",
    "    'mutation': {\n",
    "        'index': ['APC_MUT', 'CTNNB1_MUT'],\n",
    "        'alias': ['APC mut', 'CTNNB1 mut']\n",
    "    },\n",
    "    'pathway_expression': {\n",
    "        'index': ['BCAT_GDS748'],\n",
    "        'alias': ['beta-catenin activation']\n",
    "    },\n",
    "    'gene_dependency': {\n",
    "        'index': ['CTNNB1'],\n",
    "        'alias': ['CTNNB1']\n",
    "    }\n",
    "}\n",
    "\n",
    "# Load annotation bundle\n",
    "annotation_bundle = ccal.support.file.load_data_table(\n",
    "    '../data/annotation.data_table.txt', annotation_features)\n",
    "\n",
    "# Load annotation files\n",
    "annotation_files = {\n",
    "    'mutation':\n",
    "    '../results/component_annotation/c3_vs_mutation.txt',\n",
    "    'pathway_expression':\n",
    "    '../results/component_annotation/c3_vs_pathway_expression.txt',\n",
    "    'gene_dependency':\n",
    "    '../results/component_annotation/c3_vs_gene_dependency.txt'\n",
    "}\n",
    "\n",
    "# Make summary panels\n",
    "ccal.association.make_association_summary_panel(\n",
    "    target,\n",
    "    annotation_bundle,\n",
    "    annotation_files,\n",
    "    order=['mutation', 'pathway_expression', 'gene_dependency'],\n",
    "    title='Selected WNT-Related Features for KRAS Component C3',\n",
    "    filepath='../results/C3.WNT.vignette.pdf')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 4.2  Selected associations for component C6\n",
    "\n",
    "Component C6 is associated with other known alteration downstream of KRAS, the BRAF/MAPK pathway.  In this case, BRAF mutation status was the top hit associated with component C6 out of 37,276 genomic alterations. As can be seen below the component is also strongly associated with BRAF V600E and ETV1 activation signatures. ETV1 is a well-established transcription factor, downstream of the MAPK pathway, and further suggests that C6 indeed reflects a transcriptional program associated with MAPK activation. The heatmap below also shows the association of the component with the sensitivity profiles for 3 MAPK pathway inhibitors (PLX4720, PD318088 and selumetinib)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true,
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "# Select component\n",
    "component = 'C6'\n",
    "target = h_matrix.ix[component, :]\n",
    "target.name = 'KRAS Component C6'\n",
    "\n",
    "# Select specific annotation features\n",
    "annotation_features = {\n",
    "    'mutation': {\n",
    "        'index': ['BRAF.V600E_MUT'],\n",
    "        'alias': ['BRAF V600E']\n",
    "    },\n",
    "    'pathway_expression': {\n",
    "        'index': ['BRAF_UP', 'ETV1_UP'],\n",
    "        'alias': ['BRAF Oncogenic Signature', 'ETV1 Oncogenic Signature']\n",
    "    },\n",
    "    'drug_sensitivity': {\n",
    "        'index': ['PLX-4720', 'selumetinib', 'PD318088'],\n",
    "        'alias': [\n",
    "            'PLX4720 (BRAF Inhibitor)',\n",
    "            'Selumetinib (MEK1 and MEK2 Inhibitor)',\n",
    "            'PD318088 (MEK1 and MEK2 Inhibitor)'\n",
    "        ]\n",
    "    }\n",
    "}\n",
    "\n",
    "# Load annotaion bundle\n",
    "annotation_bundle = ccal.support.file.load_data_table(\n",
    "    '../data/annotation.data_table.txt', annotation_features)\n",
    "\n",
    "# Load annotation files\n",
    "annotation_files = {\n",
    "    'mutation':\n",
    "    '../results/component_annotation/c6_vs_mutation.txt',\n",
    "    'pathway_expression':\n",
    "    '../results/component_annotation/c6_vs_pathway_expression.txt',\n",
    "    'drug_sensitivity':\n",
    "    '../results/component_annotation/c6_vs_drug_sensitivity.txt'\n",
    "}\n",
    "\n",
    "# Make summary panels\n",
    "ccal.association.make_association_summary_panel(\n",
    "    target,\n",
    "    annotation_bundle,\n",
    "    annotation_files,\n",
    "    title='Selected Features for KRAS Component C6',\n",
    "    filepath='../results/C6.vignette.pdf')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 4.3  Selected associations for component C7\n",
    "\n",
    "The component C7 is significantly associated with features representing NF-κB,  a well-established pathway downstream of KRAS. This is consistent with the results of our earlier studies of KRAS synthetic lethality ([*Barbie et al. 2009*](https://www.ncbi.nlm.nih.gov/pubmed/19847166)),and RAS-driven cytokine autocrine circuits ([*Zhu et al.* 2014](https://www.ncbi.nlm.nih.gov/pubmed/24444711)).\n",
    "\n",
    "\n",
    "We show below the profiles of a gene set representing the NF-κB  motif,  an independent gene set representing p50/p65 and a profile of NF-κB protein expression. Among the genes most significantly associated with C7 was FOSL1 (FRA1) , a member of the AP-1 transcription factor family. We also observed high association of a gene set representing the targets of AP1 and the profile of the protein FRA1 pS265."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Select component\n",
    "component = 'C7'\n",
    "target = h_matrix.ix[component, :]\n",
    "target.name = 'KRAS Component C7'\n",
    "\n",
    "# Select specific annotation features\n",
    "annotation_features = {\n",
    "    'regulator': {\n",
    "        'index': ['GGGNNTTTCC_V$NFKB_Q6_01', 'V$AP1_Q4'],\n",
    "        'alias': ['NFKB TF Targets', 'AP1 TF Targets']\n",
    "    },\n",
    "    'pathway_expression': {\n",
    "        'index': ['HINATA_NFKB_TARGETS_FIBROBLAST_UP'],\n",
    "        'alias': ['Genes Up-Regulated by p50 and p65']\n",
    "    },\n",
    "    'protein_expression': {\n",
    "        'index': ['NF-kB-p65_pS536-R-C', 'FRA1_pS265-R-E'],\n",
    "        'alias': ['NF-kB p65 pS536', 'FRA1 pS265']\n",
    "    },\n",
    "    'gene_expression': {\n",
    "        'index': ['FOSL1'],\n",
    "        'alias': ['FOSL1']\n",
    "    }\n",
    "}\n",
    "\n",
    "# Load annotation bundle\n",
    "annotation_bundle = ccal.support.file.load_data_table(\n",
    "    '../data/annotation.data_table.txt', annotation_features)\n",
    "\n",
    "# Load annotation files\n",
    "annotation_files = {\n",
    "    'regulator':\n",
    "    '../results/component_annotation/c7_vs_regulator.txt',\n",
    "    'pathway_expression':\n",
    "    '../results/component_annotation/c7_vs_pathway_expression.txt',\n",
    "    'protein_expression':\n",
    "    '../results/component_annotation/c7_vs_protein_expression.txt',\n",
    "    'gene_expression':\n",
    "    '../results/component_annotation/c7_vs_gene_expression.txt'\n",
    "}\n",
    "\n",
    "# Make summary panels\n",
    "ccal.association.make_association_summary_panel(\n",
    "    target,\n",
    "    annotation_bundle,\n",
    "    annotation_files,\n",
    "    title='Selected Features for KRAS Component C7',\n",
    "    filepath='../results/C7.vignette.pdf')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 4.4  Selected associations for component C4 \n",
    "\n",
    "As can be seen below component C4 appears to be associated with ZEB1/EMT."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Select component\n",
    "component = 'C4'\n",
    "target = h_matrix.ix[component, :]\n",
    "target.name = 'KRAS Component C4'\n",
    "\n",
    "# Select specific annotation features\n",
    "annotation_features = {\n",
    "    'regulator': {\n",
    "        'index': ['V$AREB6_03', 'IPA_ZEB1'],\n",
    "        'alias': ['Targets of TCF8', 'Targets of ZEB1']\n",
    "    },\n",
    "    'pathway_expression': {\n",
    "        'index': ['TAUBE_EMT_UP', 'GROGER_EMT_UP'],\n",
    "        'alias': ['EMT Inducing TFs', 'EMT Core Gene Set']\n",
    "    },\n",
    "    'protein_expression': {\n",
    "        'index': ['N-Cadherin-R-V', 'E-Cadherin-R-V'],\n",
    "        'alias': ['N-Cadherin', 'E-Cadherin']\n",
    "    }\n",
    "}\n",
    "\n",
    "# Load annotation bundle\n",
    "annotation_bundle = ccal.support.file.load_data_table(\n",
    "    '../data/annotation.data_table.txt', annotation_features)\n",
    "# Load annotation files\n",
    "annotation_files = {\n",
    "    'regulator':\n",
    "    '../results/component_annotation/c4_vs_regulator.txt',\n",
    "    'pathway_expression':\n",
    "    '../results/component_annotation/c4_vs_pathway_expression.txt',\n",
    "    'protein_expression':\n",
    "    '../results/component_annotation/c4_vs_protein_expression.txt'\n",
    "}\n",
    "\n",
    "# Make summary panels\n",
    "ccal.association.make_association_summary_panel(\n",
    "    target,\n",
    "    annotation_bundle,\n",
    "    annotation_files,\n",
    "    title='Selected Features for KRAS Component C4',\n",
    "    filepath='../results/C4.vignette.pdf')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 4.5  Selected associations for component C2\n",
    "\n",
    "Component C2 is associated with MYC/E2F activation."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Select component\n",
    "component = 'C2'\n",
    "target = h_matrix.ix[component, :]\n",
    "target.name = 'KRAS Component C2'\n",
    "\n",
    "# Select specific annotation features\n",
    "annotation_features = {\n",
    "    'regulator': {\n",
    "        'index': ['V$E2F_02', 'V$MAX_01', 'V$MYCMAX_01', 'IPA_MYC'],\n",
    "        'alias': [\n",
    "            'Targets of E2F', 'Targets of MAX', 'Targets of MYC and MAX',\n",
    "            'Targets of MYC'\n",
    "        ]\n",
    "    }\n",
    "}\n",
    "\n",
    "# Load annotaion bundle\n",
    "annotation_bundle = ccal.support.file.load_data_table(\n",
    "    '../data/annotation.data_table.txt', annotation_features)\n",
    "\n",
    "# Load annotation files\n",
    "annotation_files = {\n",
    "    'regulator': '../results/component_annotation/c2_vs_regulator.txt',\n",
    "}\n",
    "\n",
    "# Make summary panels\n",
    "ccal.association.make_association_summary_panel(\n",
    "    target,\n",
    "    annotation_bundle,\n",
    "    annotation_files,\n",
    "    title='Selected Features for KRAS Component C2',\n",
    "    filepath='../results/C2.vignette.pdf')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 4.6  Selected associations for component C5\n",
    "\n",
    "Component C5 is associated with patterns of overexpression and dependency of HNF1 and PAX8 across multiple cancer types including subsets of ovary [*Cheung et al. 2011*](https://www.ncbi.nlm.nih.gov/pubmed/21746896), kidney, endometrial and liver."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Select component\n",
    "component = 'C5'\n",
    "target = h_matrix.ix[component, :]\n",
    "target.name = 'KRAS Component C5'\n",
    "\n",
    "# Select specific annotation features\n",
    "annotation_features = {\n",
    "    'gene_expression': {\n",
    "        'index': ['PAX8', 'HNF1B'],\n",
    "        'alias': ['PAX8', 'HNF1B']\n",
    "    },\n",
    "    'gene_dependency': {\n",
    "        'index': ['PAX8', 'HNF1B'],\n",
    "        'alias': ['PAX8', 'HNF1B']\n",
    "    },\n",
    "    'tissue': {\n",
    "        'index': ['kidney', 'ovary', 'endometrium', 'liver'],\n",
    "        'alias': ['Kidney', 'Ovary', 'Endometrium', 'Liver']\n",
    "    }\n",
    "}\n",
    "\n",
    "# Load annotaion bundle\n",
    "annotation_bundle = ccal.support.file.load_data_table(\n",
    "    '../data/annotation.data_table.txt', annotation_features)\n",
    "\n",
    "# Load annotation files\n",
    "annotation_files = {\n",
    "    'gene_expression':\n",
    "    '../results/component_annotation/c5_vs_gene_expression.txt',\n",
    "    'gene_dependency':\n",
    "    '../results/component_annotation/c5_vs_gene_dependency.txt',\n",
    "    'tissue':\n",
    "    '../results/component_annotation/c5_vs_tissue.txt',\n",
    "}\n",
    "\n",
    "# Make summary panels\n",
    "ccal.association.make_association_summary_panel(\n",
    "    target,\n",
    "    annotation_bundle,\n",
    "    annotation_files,\n",
    "    title='Selected Features for KRAS Component C5',\n",
    "    filepath='../results/C5.vignette.pdf')"
   ]
  }
 ],
 "metadata": {
  "anaconda-cloud": {},
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
